applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -135,38 +144,94 @@ static void handle_dvfs_error(int32_t err)
 	} else if (err == -EAGAIN) {
 		LOG_DBG("DVFS not initialized, trying again.");
 		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
+	} else if (err == -ETIMEDOUT) {
+		LOG_DBG("DVFS frequency change timed out, trying again.");
+		timeout = DVFS_RETRY_BUSY_TIMEOUT;
 	} else {
 		LOG_ERR("DVFS freq change returned with error: %d", err);
-		module_set_state(MODULE_STATE_ERROR);
-		module_state = STATE_ERROR;
-		cancel_dvfs_retry_work();
+		dvfs_fatal_error();
 		return;
 	}
 	(void) k_work_reschedule(&dvfs_retry.retry_work, timeout);
 }
 
-static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+static uint32_t check_required_frequency(void)
+{
+	if (dfvs_requests_state_bitmask & dvfs_high_freq_bitmask) {
+		return CONFIG_DESKTOP_DVFS_FREQ_HIGH;
+	} else if (dfvs_requests_state_bitmask & dvfs_medlow_freq_bitmask) {
+		return CONFIG_DESKTOP_DVFS_FREQ_MED;
+	} else {
+		return CONFIG_DESKTOP_DVFS_FREQ_LOW;
+	}
+}
+
+static void dvfs_notify_cb(void)
+{
+	(void) k_work_submit(&dvfs_notify_work);
+}
+
+static void set_dvfs_freq(uint32_t target_freq)
 {
-	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+	int ret;
+
+	if (spec.frequency != 0) {
+		ret = nrf_clock_control_release(dvfs_clock_dev, &spec);
+		if (ret < 0) {
+			LOG_ERR("Failed to release requested clock specs, error: %d", ret);
+			dvfs_fatal_error();
+			return;
+		}
+	}
 
+	sys_notify_init_callback(&cli.notify, dvfs_notify_cb);
+
+	spec.frequency = target_freq;
+	ret = nrf_clock_control_request(dvfs_clock_dev, &spec, &cli);
 	if (ret) {
 		handle_dvfs_error(ret);
 	} else {
-		current_freq = target_freq;
-		LOG_INF("Have requested %s frequency",
-			get_dvfs_frequency_setting_name(target_freq));
-		cancel_dvfs_retry_work();
+		LOG_INF("Have requested %" PRIu32 " frequency", target_freq);
+		requested_freq = target_freq;
+		request_in_progress = true;
 	}
 }
 
-static enum dvfs_frequency_setting check_required_frequency(void)
+static void dvfs_notify_work_handler(struct k_work *work)
 {
-	if (dfvs_requests_state_bitmask & dvfs_high_freq_bitmask) {
-		return DVFS_FREQ_HIGH;
-	} else if (dfvs_requests_state_bitmask & dvfs_medlow_freq_bitmask) {
-		return DVFS_FREQ_MEDLOW;
-	} else {
-		return DVFS_FREQ_LOW;
+	int res;
+	int ret = sys_notify_fetch_result(&cli.notify, &res);
+
+	if (ret < 0) {
+		LOG_WRN("Work not completed, verify usage of notify API");
+		return;
+	}
+
+	request_in_progress = false;
+
+	if (res < 0) {
+		handle_dvfs_error(res);
+		return;
+	}
+	ret = clock_control_get_rate(dvfs_clock_dev, NULL, &current_freq);
+	if (ret < 0) {
+		LOG_ERR("Failed to get current frequency with error: %d", ret);
+		dvfs_fatal_error();
+		return;
+	}
+	__ASSERT(requested_freq == current_freq,
```

Comment:
Why fatal error is better then assert? 

If there is a different module which will use dvfs directly - dvfs will accept this request and handle it. 
If other module uses clock control api then I suppose it depends how it will be used - but I guess another instance of cli will be created and this two api will not see each other and won't negotiate dvfs oppoint between themselves.
So there still will be a mess.

Reply:
Fatal error would also trigger in release configurations (and it would allow us to see the "mess" clearly). Currently we just emit warning log and retry in (potentially) never ending loop.

Reply:
I understand that our dvfs module could not negotiate, but it would be worth to check if we shouldn't e.g. accept frequencies higher than requested then.

Reply:
Discussed offline. We will emit fatal error for now and later on improve the implementation to properly handle case with multiple users of the clock control API